### DISCUSSION TITLE
Error when building the package using Colcon (mixin)

### DISCUSSION THREAD

**Discussion by axelssons** [Yaskawa-Global/motoros2_client_interface_dependencies#15] [Q&A] [answered]
While precisely following the installation and build instructions in the README-file of this repository, I get an error while running:
`colcon build --mixin release --packages-up-to motoros2_client_interface_dependencies`

The error is:
`colcon: error: Mixin 'release' is not available for 'build'`

Why do I get this and is this troubling? When running just `colcon build` it works fine. Will it work this way or do I have to build it with the --mixin and --packages-up-to flags?

**Comment by gavanderhoorn** [comment, marked as answer]
> The error is: `colcon: error: Mixin 'release' is not available for 'build'`

it looks like you don't have `colcon_mixin` installed.

The ROS 2 Docker images (ie: `ros:humble`, etc) come with that installed by default, which is perhaps why our tests didn't show that as missing.

> Why do I get this [..]? When running just `colcon build` it works fine. Will it work this way or do I have to build it with the `--mixin` and `--packages-up-to` flags?

The main difference would be:

- all packages in your workspace get built, instead of just `motoros2_client_interface_dependencies` and its dependencies, and
- your packages are built with `Debug` build type, instead of the `Release` build type, meaning no compiler optimisation will be enabled

> and is this troubling

it isn't really problematic, if that's what you're asking.

Building with `Release` build type enabled will typically make binaries more performant.

In most cases however, you'll not notice it (your HW is likely 'fast enough' to compensate).

**Reply by gavanderhoorn** [reply]
If you have ROS already installed (and thus the package repositories enabled), you could install `python3-colcon-mixin` and then you should be able to build using `--mixin release`.

Note that `--packages-up-to` is basic Colcon functionality and is not related to `colcon_mixin`.

**Reply by axelssons** [reply]
Thank you!

Apparently, `colcon_mixin` was already installed. However, I needed to register [the colcon mixin repository ](https://github.com/colcon/colcon-mixin-repository) with: 
`colcon mixin add default https://raw.githubusercontent.com/colcon/colcon-mixin-repository/master/index.yaml` 
followed by `colcon mixin update default`.
